### ISSUE TITLE
Proper way to zero out padding elements

### ISSUE THREAD

**Issue by sooheon** [OliverRichter/normalized-attention#1]
In softmax attention, to avoid attending to padding elements, you add a -inf value to padded positions so their softmax value becomes 0.

For normalized attention, there is no simple equivalent operation without adding a lot of complexity.

**Before norm**: won't work, the recentering and rescaling means that your zeroed out padding will not remain zero.
**After norm**: you can zero out the post-norm attention weights, but this means your remaining elements are not zero-mean unit variance.

Only way I can think of to get the desired effect is to calculate the mean-std for normalization with taking the padding elements into account, but I'm not sure how to achieve this with GPU vectorization.

**Comment by OliverRichter** [comment]
Hi,
(Sorry for the delayed reply, I'm currently on vacation)
You're right, there is no easy hack as in the softmax attention that would allow for vectorization through padding. In this repo I decided to allow for a dynamic input sequence length (no padding required). While this might lack the speed up you gain from vectorization, it does not require a 'max-sequence-length' hyperparameter* and avoids the calculation overhead introduced by the padding on short sequences. As one of the arguments for normalized attention is its independence to the sequence length (as compared to softmax attention, see paper) I opted for this dynamic implementation, as it allows for input sequences of vastly different lengths.
If you do wish to have a version that allows for padding/vectorization you can write a normalization function that accepts special padding tokens (like 'None') in the input sequence. I might include this in the future.
Let me know if you have any further question.

*The 'max-sequence-lenght' hyperparameter in the code is only used for the positional embedding, not in the attention

**Comment by sooheon** [comment]
While implementing this, I noticed that if the sequence length happens to be 1, the returned normalized attention weight for that single element must be 0 (zero mean, unit variance). Of course for softmax, a single element will always be weighted at 1. What do you think is the "correct" way to deal with this edge case? I guess allowing the attention weight to be 0 is at least justifiable if we consider that the skip-connection will still pass information.

**Comment by OliverRichter** [comment]
In this edge case, the bias term* can allow for an information flow (even without considering the skip connection). However, if you want to have the property that the magnitude of the vector output is independent of the sequence length even in this case, you can return a 1 as output of the normalization function in this edge case. I did not study this in depth, but returning a 1 here might yield better results (depending on your application), as the network might propagate the training signal better early on in training. But this is just a guess. Let me know if you gain any insights here. 

*the normalized attention weights are calculated as attention = gain * normalize(logits) + bias

**Comment by sooheon** [comment]
For refrence, here's my very naive pytorch implementation. Rather than take special tokens, it takes an explicit mask denoting which positions to ignore.

```python
def masked_normalize(x, float_mask, dim=-1, eps=1e-6):
    """
    Normalize tensor along dim to zero mean unit variance, but without considering
    positions indicated as 0 in float_mask.

    :param x: tensor to normalize.
    :param float_mask: Mask of same shape as x. 0 for removal, 1 for pass through.
    :param dim: Dimension to normalize. Usually -1, which contain features or pooling
      weights.
    :param eps: Small clamp value to avoid division by zero.
    :return: Normalized tensor with mask applied before normalization.
    """
    counts = torch.sum(float_mask, dim=dim, keepdim=True)
    means = torch.sum(x * float_mask, dim=dim, keepdim=True) / counts
    errors = (x - means) * float_mask
    variances = torch.sum(torch.pow(errors, 2), dim=dim, keepdim=True) / counts
    stddevs = variances.sqrt()
    normalized = errors / stddevs.clamp_min(eps)
    return normalized
```

**Comment by OliverRichter** [comment]
Nice, thanks for sharing!
